ICSharpCode.Decompiler/IL/Transforms/InlineArrayTransform.cs

```diff
@@ -0,0 +1,234 @@
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+using ICSharpCode.Decompiler.TypeSystem;
+
+namespace ICSharpCode.Decompiler.IL.Transforms
+{
+	static class InlineArrayTransform
+	{
+		internal static bool RunOnExpression(Call inst, StatementTransformContext context)
+		{
+			if (MatchSpanIndexerWithInlineArrayAsSpan(inst, out var type, out var addr, out var index, out bool isReadOnly))
+			{
+				if (isReadOnly)
+				{
+					context.Step("call get_Item(addressof System.ReadOnlySpan{T}(call InlineArrayAsReadOnlySpan(addr)), index) -> readonly.ldelema.inlinearray(addr, index)", inst);
+				}
+				else
+				{
+					context.Step("call get_Item(addressof System.Span{T}(call InlineArrayAsSpan(addr)), index) -> ldelema.inlinearray(addr, index)", inst);
+				}
+				inst.ReplaceWith(new LdElemaInlineArray(type, addr, index) { IsReadOnly = isReadOnly }.WithILRange(inst));
+				return true;
+			}
+
+			if (MatchInlineArrayElementRef(inst, out type, out addr, out index, out isReadOnly))
+			{
+				if (isReadOnly)
+				{
+					context.Step("call InlineArrayElementRefReadOnly(addr, index) -> readonly.ldelema.inlinearray(addr, index)", inst);
+				}
+				else
+				{
+					context.Step("call InlineArrayElementRef(addr, index) -> ldelema.inlinearray(addr, index)", inst);
+				}
+				inst.ReplaceWith(new LdElemaInlineArray(type, addr, index) { IsReadOnly = isReadOnly }.WithILRange(inst));
+				return true;
+			}
+
+			if (MatchInlineArrayFirstElementRef(inst, out type, out addr, out isReadOnly))
+			{
+				if (isReadOnly)
+				{
+					context.Step("call InlineArrayFirstElementRefReadOnly(addr) -> readonly.ldelema.inlinearray(addr, ldc.i4 0)", inst);
+				}
+				else
+				{
+					context.Step("call InlineArrayFirstElementRef(addr) -> ldelema.inlinearray(addr, ldc.i4 0)", inst);
+				}
+				inst.ReplaceWith(new LdElemaInlineArray(type, addr, new LdcI4(0)) { IsReadOnly = isReadOnly }.WithILRange(inst));
+				return true;
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// Matches call get_Item(addressof System.(ReadOnly)Span[[T]](call InlineArrayAs(ReadOnly)Span(addr, length)), index)
+		/// </summary>
+		static bool MatchSpanIndexerWithInlineArrayAsSpan(Call inst, [NotNullWhen(true)] out IType? type, [NotNullWhen(true)] out ILInstruction? addr, [NotNullWhen(true)] out ILInstruction? index, out bool isReadOnly)
+		{
+			isReadOnly = false;
+			type = null;
+			addr = null;
+			index = null;
+
+			if (MatchSpanGetItem(inst.Method, "ReadOnlySpan"))
+			{
+				isReadOnly = true;
+
+				if (inst.Arguments is not [AddressOf { Value: Call targetInst, Type: var typeInfo }, var indexInst])
+					return false;
+
+				if (!MatchInlineArrayHelper(targetInst.Method, "InlineArrayAsReadOnlySpan", out var inlineArrayType))
+					return false;
+
+				if (targetInst.Arguments is not [var addrInst, LdcI4])
```

Comment:
The value of the LdcI4 needs to be checked against the length of the inline array type.